### PR TITLE
Add support for problem-specific sources in N Euler

### PR DESCRIPTION
**Pull request by carmaza**
Some tests in Newtonian hydro require problem-specific sources. This PR adds support for that. As a simple example, this PR adds the sources generated by a uniform acceleration (e.g. constant gravity), which are useful when studying non-linear instabilities (e.g Rayleigh-Taylor inst.). Other sources to be added in the future include central object potentials and "stirring" functions that drive turbulence.

- [ ] Bugfix
- [x ] New feature

### Component:

- [x ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.cpp
Comment (wthrowe) [inline src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.cpp]: const

File: src/Evolution/Systems/NewtonianEuler/Sources.hpp
Comment (wthrowe) [inline src/Evolution/Systems/NewtonianEuler/Sources.hpp]: Is `Args&` intended here, or should it be `const Args&` or `Args&&`?

**Review by wthrowe — CHANGES_REQUESTED** [review]
Does it make sense to "mix and match" sources and analytic data/solutions, or do the values in the metavariables have to be kept in sync?  if the latter, could the sources be specified by the analytic data class?

**Comment by carmaza** [comment]
@wthrowe Ideally, we'd like that any parameter(s) shared both by analytic data and sources be available from a common variable. I'm open for suggestions on how to do that!

File: src/Evolution/Systems/NewtonianEuler/Sources.hpp
Comment (carmaza) [reply src/Evolution/Systems/NewtonianEuler/Sources.hpp]: I think here I just chose the one that wasn't giving compilation errors. Which one should be the preferred here?

Comment (wthrowe) [reply src/Evolution/Systems/NewtonianEuler/Sources.hpp]: `const Args&` should be correct if it compiles.  If that doesn't work I'll take a closer look at exactly how this is being used.

**Comment by wthrowe** [comment]
The simplest way I can think of would be to have a method on the analytic data that returns the appropriate sources object and then use the new flexible OptionTag -> Tag mapping to construct both from the same input file section.

File: src/Evolution/Systems/NewtonianEuler/Sources.hpp
Comment (carmaza) [reply src/Evolution/Systems/NewtonianEuler/Sources.hpp]: `const Args&` worked. Not sure if/why I tried that before :P

**Comment by carmaza** [comment]
Regard this as a whole new PR. Commit order should be:

- Add UniformAcceleration source for NewtonianEuler
- Mark IsentropicVortex constructor noexcept
- Add function for computing perturbation term in IsentropicVortex
- Rename IsentropicVortexSource VortexPerturbation
- Update VortexPerturbation source to evolve modified IsentropicVortex
- Add source term member to IsentropicVortex
- Force 2D IsentropicVortex to use vanishing perturbation
- Add NoSource type for NewtonianEuler
- Add source_term_type alias in RiemannProblem
- Add ComputeSources for NewtonianEuler
- Add Tags for problem-specific sources for NewtonianEuler
- Add Initialization of Source Term for NewtonianEuler
- Enable source terms in NewtonianEuler

File: src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.hpp
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.hpp]: `NOLINT(readability-redundant-declaration)` should no lint this only for the specified argument

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp]: `function_of_z` -> `perturbation_profile`

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp]: `deriv_of_perturbation_profile`

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp
Comment (kidder) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp]: `z` -> `z_coord`

File: src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp]: `2-D` -> `2d` and similarly for `3d`

File: src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp]: `NOLINT(rereadability-redundant-declaration)`

File: src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp]: Instead of creating the solution every time `apply` is called, just pass it to `apply` and then you don't need these member variables

File: src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp]: pass the solution as an argument so it is not recomputed everytime... (see comment in hpp file)

File: src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp]: `size` -> `number_of_grid_points`

File: src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp]: save an allocation by storing `dz_vortex_velocity_z` in `*source_mass_density_cons`, and then do the multiplication by `vortex_mass_density_cons` after computing the other sources

**Review by kidder — CHANGES_REQUESTED** [review]
partial review with @nilsdeppe , got through `add source term member to IsentropicVortex`

**Comment by carmaza** [comment]
Thanks for starting reviewing this. I'll push the fixup when the review is done. In the meantime, I'm pushing an extra commit to add some `__init__.py` files to fix a test that was broken. (The absence of these files will only break runs using python 2)

File: src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp
Comment (wthrowe) [inline src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.hpp]: Can these be `tmpl::conditional_t<Dim == 3, tmpl::list<stuff>, tmpl::list<>>`?

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.cpp
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.cpp:42]: You can enforce this when constructed from an input file by only including `PerturbAmplitude` in the `options` list in 3D.  You'll have to overload the constructor or reorder the arguments and default `perturbation_amplitude`.  (If that doesn't work for some reason this should be a `PARSE_ERROR` because it can be triggered by an input file.)

The assertion is still good to have in case someone creates the class in C++.

File: src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp
Comment (wthrowe) [inline src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp]: You don't use the initial time.  I don't think you need this Requires.

File: src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp
Comment (wthrowe) [inline src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp]: Would it make sense to have the source term in the cache along with the analytic solution?

File: src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp
Comment (wthrowe) [inline src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp]: `cpp17::is_same_v<typename Metavariables::source_term_type, NewtonianEuler::Sources::NoSource>`, and use `tmpl::conditional_t`.  Again below.  (Same as is done with the local time stepping stuff.)

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: tests/Unit/Evolution/Systems/NewtonianEuler/Test_Sources.cpp
Comment (kidder) [inline tests/Unit/Evolution/Systems/NewtonianEuler/Test_Sources.cpp]: since all the bounds are the same, you can just pass a single `{-1.0, 1.0}` and change the `4` on line 144 to `1`

File: tests/Unit/Evolution/Systems/NewtonianEuler/Test_Sources.cpp
Comment (kidder) [inline tests/Unit/Evolution/Systems/NewtonianEuler/Test_Sources.cpp]: again a single bound

File: src/Evolution/Systems/NewtonianEuler/Tags.hpp
Comment (kidder) [inline src/Evolution/Systems/NewtonianEuler/Tags.hpp:111]: add a test for  `name`

File: src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp
Comment (kidder) [reply src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp]: Yes, I think that makes sense

**Review by kidder — CHANGES_REQUESTED** [review]

**Comment by kidder** [comment]
@wthrowe please look at the fixups

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp
Comment (wthrowe) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp]: Remove PerturbAmplitude in 2d.

File: src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp
Comment (wthrowe) [reply src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp]: ping

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp
Comment (carmaza) [reply src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp]: Would using a `tmpl::conditional_t` suffice?

File: src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp
Comment (carmaza) [reply src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp]: Most of the ideas devoted to this change would require some changes beyond the scope of this PR, and they would also be applicable to putting EoSs (and possibly other classes) in the global cache. These changes would certainly imply further group discussion, so, after discussing this with @kidder, I opted to leave the source in the databox for now and come back with an enhancement once we know how we want to do it with EoSs.

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp
Comment (wthrowe) [reply src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp]: Yes.

File: src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp
Comment (wthrowe) [reply src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp]: Is there more needed for this than
```c++
template <typename SolutionType>
struct SourceTerm : SourceTermBase, db::SimpleTag {
  using type = typename SolutionType::source_term_type;
  using option_tags = tmpl::list<::OptionTags::AnalyticSolution<SolutionType>>;
  static type create_from_options(
      const SolutionType& analytic_solution) noexcept {
    return analytic_solution.sources();
  }
};
```
and changing some template parameters around?

Comment (nilsdeppe) [reply src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp]: Yea, I think that should work. Maybe we should talk about how to implement this after the exec call? I think EoSs could be done the same or in a similar way.

Comment (carmaza) [reply src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp]: That was my first attempt, but the initial data is not restricted to analytic solutions, or even to analytic data. Adding an `OptionTags` for a generic `InitialData` type might do it, but the very fact of creating the source (or the EoS) as a member of another class seems ultimately suboptimal in cases where you don't really need it to be a member (i.e., when the source doesn't depend on parameters that needed for the initial data, e.g. for numerical initial data.) I'm fine with implementing your suggestion if you think that it will make things easier to update in the future, though.

Comment (carmaza) [reply src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp]: I'd be happy to discuss what's required for (and volunteer to implement) the changes for EoSs.

Comment (carmaza) [reply src/Evolution/Systems/NewtonianEuler/InitializeSourceTerm.hpp]: So after our discussion today, I implemented Will's suggestion (last fixup commit). The initialization source term commit will be dropped once I squash.

File: src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp
Comment (nilsvu) [inline src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp]: you can remove the `flatten` as of #1699

File: src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.hpp
Comment (nilsvu) [inline src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.hpp:65]: If you rename `sourced_variables` -> `return_tags` and `db::wrap_tags_in<Tags::Source, ...>` then you don't need the `ComputeSources` wrapper in the other file. I don't think it's worth having that wrapper only to do the `wrap_tags_in`. The system can then directly retrieve `volume_sources = typename InitialDataType::sources`.

File: src/Evolution/Systems/NewtonianEuler/Tags.hpp
Comment (nilsvu) [inline src/Evolution/Systems/NewtonianEuler/Tags.hpp:100]: In the elliptic code I have an `elliptic::Tags::FluxesComputer`, so you could name this tag `SourcesComputer`. I don't think we have naming guidelines for these sort of tags yet, but we can try to be consistent.

File: src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp
Comment (nilsvu) [inline src/PointwiseFunctions/AnalyticSolutions/NewtonianEuler/IsentropicVortex.hpp:97]: I think the simpler `using sources = Sources::VortexPerturbation<Dim>` is more in line with how we name these aliases e.g. in the `Metavariables`. No need to add a `_type` suffix (could also be removed from the EOS in the line above, but not in this PR).

File: tests/Unit/Evolution/Systems/NewtonianEuler/Test_Sources.cpp
Comment (nilsvu) [inline tests/Unit/Evolution/Systems/NewtonianEuler/Test_Sources.cpp:158]: If I understand correctly, this test is only for the `ComputeSources` class, which does nothing but wrap the `sourced_variables` in `Tags::Sources`. If you do that in the source class directly (see comment above about `sourced_variables` -> `return_tags`) then you don't need this test at all.

Even if you were to keep this test, I don't think you should go through a Python implementation of the sources that are _only used within this test_. Unit tests should test no more and no less than the piece of code they are designed to test. In this case, that would only be that `ComputeSources<...>` is compatible with `mutate_apply` and that it returns `Tags::Source<...>`.

**Review by nilsvu — COMMENTED** [review]
Not a complete review, but things I noticed when reading through this PR. Looks like this PR has enough reviewers already, so don't wait for my approval for fixups or merging.

File: src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.hpp
Comment (carmaza) [reply src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.hpp:65]: Considering that our generic `ComputeVolumeSources` action for evolution systems is a wrapper to the databox function `mutate_apply`, I am not sure how to make the latter work with stateful structs such as sources holding their own members.

Comment (nilsvu) [reply src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.hpp:65]: Ah, I see now that your `ComputeSources` wrapper also adds the `Tags::SourceTerm` to the `argument_tags` and forwards the apply call to it as a member function.
Note that `mutate_apply` also has an overload for stateful structs, i.e. the following would work:
```cpp
db::mutate_apply(get<Tags::SourceTerm<InitialDataType>>(box), make_not_null(&box))
```
We could actually specialize the stateless `db::mutate_apply<Invokable>(make_not_null(&box))` overload for the case where `Invokable` is a tag, such that it retrieves the tag and calls the stateful overload automatically. But this is out of scope for this PR, so having the `ComputeSources` wrapper seems like a good solution to me.

Comment (nilsvu) [reply src/Evolution/Systems/NewtonianEuler/Sources/UniformAcceleration.hpp:65]: I pushed an implementation of such a specialization to #1728

**Comment by wthrowe** [comment]
I am happy with the fixups for my comments.

**Comment by kidder** [comment]
I am happy with the fixups as well

**Comment by carmaza** [comment]
Well, I'm ready to squash whenever.

File: src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp
Comment (nilsdeppe) [reply src/Evolution/Executables/NewtonianEuler/EvolveNewtonianEuler.hpp]: ping on this.

**Review by nilsdeppe — COMMENTED** [review]
I'm happy for you to squash.

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
@wthrowe @nilsdeppe this has been squashed and rebased

**Review by nilsdeppe — APPROVED** [review]

**Comment by kidder** [comment]
@wthrowe if you are happy with this, please merge it

File: src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp
Comment (wthrowe) [inline src/Evolution/Systems/NewtonianEuler/Sources/VortexPerturbation.cpp:63]: A simpler way to do this would have been to reuse the allocation returned by `deriv_of_perturbation_profile`.  (Assign it to a variable and then `*=` the amplitude.)  (Just a comment.  Not requesting a change.)

**Review by wthrowe — APPROVED** [review]
@nilsleiffischer If there are any further changes you want, please file issues for them.

Ignoring Travis timeout.

**Pull request merged**
